### PR TITLE
[refactor] Get rid of band class in favour of pure functions

### PR DESCRIPTION
**Pull request by toxa81**
* Band class is removed
* all diagonalziation functions are moved to ./hamiltonian folder' ./band folder is removed

### CODE REVIEW

**Review by simonpintarelli — APPROVED** [review]
LGTM. I've added my proposal for changes in https://github.com/electronic-structure/SIRIUS/pull/890

File: src/hamiltonian/diagonalize_pp.hpp [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+template <typename T, typename F>
+inline std::enable_if_t<!std::is_same<T, real_type<F>>::value, void>
+diagonalize_pp_exact(int ispn__, Hamiltonian_k<T>& Hk__)
+{
+    RTE_THROW("not implemented");
+}
+
+template <typename T, typename F>
+inline std::enable_if_t<std::is_same<T, real_type<F>>::value, void>
+diagonalize_pp_exact(int ispn__, Hamiltonian_k<T>& Hk__)
```
[the remark this answers is not in the document]
Comment (toxa81) [reply src/hamiltonian/diagonalize_pp.hpp:41]: Let's do it in the next PR. As this one is greenlighted, I'll merge it.

File: apps/dft_loop/sirius.scf.cpp [hunk 2]
```diff
@@ -495,17 +497,16 @@ void run_tasks(cmd_args const& args)
         //density.initial_density();
         density.load(storage_file_name);
         potential.generate(density, ctx->use_symmetry(), true);
-        Band band(*ctx);
         Hamiltonian0<double> H0(potential, true);
         if (!ctx->full_potential()) {
-            band.initialize_subspace(ks, H0);
+            ::sirius::initialize_subspace(ks, H0);
```
[the remark this answers is not in the document]
Comment (toxa81) [reply apps/dft_loop/sirius.scf.cpp]: fixed

File: python_module/py_sirius.cpp [hunk 3]
```diff
@@ -104,7 +105,7 @@ initialize_subspace(DFT_ground_state& dft_gs, Simulation_context& ctx)
 {
     auto& kset = dft_gs.k_point_set();
     Hamiltonian0<double> H0(dft_gs.potential(), false);
-    Band(ctx).initialize_subspace(kset, H0);
+    ::sirius::initialize_subspace(kset, H0);
```
[the remark this answers is not in the document]
Comment (toxa81) [reply python_module/py_sirius.cpp]: fixed

**Pull request merged**